### PR TITLE
[mlir] Start moving some builtin type formats to the dialect

### PR DESCRIPTION
**Pull request by zero9178**
Most types and attributes in the builtin dialect are parsed and printed using special-purpose printers and parsers for that type. They also use the low-level `Printer` rather than the `AsmPrinter`, making the implementations inconsistent compared to all other dialects in MLIR.

This PR starts moving some builtin types to be parsed using the usual `print` and `parse` methods like all other MLIR dialects. This has the following advantages:
* The implementation now looks like any other dialect's
* It is now possible to use `assemblyFormat` for builtin types and attributes
* The code can be easily moved to other dialects if desired
* Arguably better layering and less code
* As a side-effect, it is now also possible to write `!builtin.<type>` for any types if desired

A future benefit would include being able to print types and attributes in stripped format as well (e.g. `<f32>` vs `complex<f32>`), just like all other dialect types and attributes. This is currently explicitly disabled as it causes a LOT of changes in IR syntax and I believe some ambiguities in the parser.

For the purpose of reviewing and incremental development, this PR only moves `tuple`, `tensor`, `none`, `memref` and `complex`. The plan is to eventually move all attributes and types where the current syntax can be implemented within the dialect.

For backwards compatibility with the existing syntax, the builtin dialect is special-cased in the printer where the `builtin.` prefix is omitted.

Depends on https://github.com/llvm/llvm-project/pull/80420

### CODE REVIEW

File: mlir/include/mlir/IR/BuiltinTypes.td [hunk 1]
```diff
@@ -25,7 +25,8 @@ include "mlir/IR/BuiltinTypeInterfaces.td"
 // Base class for Builtin dialect types.
 class Builtin_Type<string name, string typeMnemonic, list<Trait> traits = [],
                    string baseCppClass = "::mlir::Type">
-    : TypeDef<Builtin_Dialect, name, traits, baseCppClass> {
+    : TypeDef<Builtin_Dialect, name, !listconcat(traits, [PrintTypeQualified]),
```
Comment (joker-eph) [inline mlir/include/mlir/IR/BuiltinTypes.td:28]: I don't quite follow why `PrintTypeQualified` here? Do you have a test which shows that?

Comment (zero9178) [reply mlir/include/mlir/IR/BuiltinTypes.td:28]: The best example as to why this is needed is probably `complex`. Not using `PrintTypeQualified` means changing the syntax of ALL operations in the complex dialect as they all suddenly use the unqualified syntax. 
Additionally, `parseCustomTypeWithFallback` currently does not handle the builtin types correctly if written verbose (e.g. `complex<f32>`) since the logic currently assumes that verbose types always lead with the dialect name (`!builtin`).
I tried a bit what logic changes would be required and did not find an obvious answer. It therefore also breaks existing input parsing.
So to summarize:
* To stay compatible with existing syntax and tests and avoid creating massive churn
* To avoid incompatiblity with the current implementation of stripped parsing and printing

If you like I can take a closer look at fixing `parseCustomTypeWithFallback` here, but thought I'd rather make that a future PR to keep this one backwards compatible.

**Comment by zero9178** [comment]
ping
